pykafka/protocol.py

```diff
@@ -794,17 +814,45 @@ def _unpack_message_set(self, buff, partition_id=-1):
         for message in message_set.messages:
             if message.compression_type == CompressionType.NONE:
                 output.append(message)
+                continue
             elif message.compression_type == CompressionType.GZIP:
                 decompressed = compression.decode_gzip(message.value)
-                output += self._unpack_message_set(decompressed,
-                                                   partition_id=partition_id)
+                messages = self._unpack_message_set(decompressed,
+                                                    partition_id=partition_id)
             elif message.compression_type == CompressionType.SNAPPY:
                 decompressed = compression.decode_snappy(message.value)
-                output += self._unpack_message_set(decompressed,
-                                                   partition_id=partition_id)
+                messages = self._unpack_message_set(decompressed,
+                                                    partition_id=partition_id)
+            if messages[-1].offset < message.offset:
+                # With protocol 1, offsets from compressed messages start at 0
+                assert messages[0].offset == 0
+                delta = message.offset - len(messages) + 1
```

Comment:
This is extremely empirical. Suppose you start at offset 5, then receive 3 compressed messages. The compressed enclosing message will have offset 8, and the enclosed messages will have offsets (6, 7, 8) in protocol 0, or (0, 1, 2) in protocol 1. So I have to to add 5, which is 8-3, to the protocol-1 offsets to get back the 6, 7, 8 offsets, which is what I would be getting with uncompressed messages. Is it clearer?